keps/20200901-standard-crd.md

```diff
@@ -0,0 +1,240 @@
+3. AS a KES user i want to control the sink for the secrets (aka frontend: store secret as `kind=ConfigMap` or `kind=Secret`)
+4. AS a KES user i want to fetch **from multiple** stores and store the secrets **in a single** Frontend
+5. AS a KES operator i want to limit the access to certain stores or subresources (e.g. having one central KES instance that handles all ES - similar to `iam.amazonaws.com/permitted` annotation per namespace)
+
+### Stores
+
+These stores are relevant:
+* AWS Secure Systems Manager Parameter Store
+* AWS Secrets Manager
+* Hashicorp Vault
+* Azure Key Vault
+* Alibaba Cloud KMS Secret Manager
+* Google Cloud Platform Secret Manager
+* Kubernetes (see #422)
+* noop (see #476)
+
+### Frontends
+
+* Kind=Secret
+* Kind=ConfigMap
+* *potentially* we could sync store to store
+
+## Proposal
+
+### API
+
+### External Secret
+
+The `ExternalSecret` CustomResourceDefinition is **namespaced**. It defines the following:
+1. source for the secret (store)
+2. sink for the secret (fronted)
+3. and a mapping to translate the keys
+
+```yaml
+apiVersion: kes.io/v1alpha1
+kind: ExternalSecret
+metadata: {...}
+
+spec:
+
+  # optional.
+  # used to select the correct KES instance (think: ingress.ingressClassName)
+  # There is no need for a indirection (e.g. having a extra resource like kind=IngressClass)
+  # the KES controller is instantiated with a specific class name
+  # and filters ES based on this property
+  className: "dev"
+
+  # the amount of time before the values will be read again from the store
+  refreshInterval: "1h"
+
+  # there can only be one frontend per ES
+  # this is the "thing" that is created by KES.
+  # conceptually speaking a frontend is just a thing that can hold k/v pairs
+  frontend:
+    secret:
+      # do we need a api version here? who handles upgrades?
+      apiVersion: v1
+      name: my-secret
+      template:
+        type: kubernetes.io/dockerconfigjson # or TLS...
+        metadata:
+          annotations: {}
+          labels: {}
+    # of course only one frontend should be defined
+    configMap:
+      # do we need a api version here? who handles upgrades?
+      apiVersion: v1 #
+      name: my-configmap
+      template:
+        metadata:
+          labels:
+            app: foo
+
+  data:
+
+    # EXAMPLE 1: simple mapping
+    # one key from a store may hold multiple values
+    # we need a way to map the values to the frontend
+    # it is the responsibility of the store implementation to know how to extract a value
+  - key: /corp.org/dev/certs/ingress
+    property: pubcert
+    name: tls.crt
+  - key: /corp.org/dev/certs/ingress
+    property: privkey
+    name: tls.key
+
+  # EXAMPLE 2: multiple stores per ES
+  # we also need a way to fetch secrets from multiple stores
+  # thus, we need a way to define a store per data key
+  - key: /rds/database-secret
+    name: database
+    storeRef:
```

Comment:
Having an ExternalSecret able to reference multiple stores complicates the ability to what secrets have been created from a specific store. This field seems better at the spec level rather than in each data reference.

Container are already able to mount any number of secrets, if data from multiple stores is needed in a single secret for another reason, the design around being able to merge ExternalSecrets data into an existing secret seems a bit clearer for this use case - https://github.com/godaddy/kubernetes-external-secrets/pull/477/files#r487474442,

Reply:
Yeah, I feel having different stores for different keys adds complexity for sure. For KES there was some feature requests for this before in https://github.com/godaddy/kubernetes-external-secrets/issues/132 
Main use case there seemed to be migration from one secret store to another. Which I'm not sure if that's enough of a reason to add the complexity? If one really is migrating from one store to another wouldn't you try to at least migrate all the keys used in the same secret. 

Using multiple external secrets to target the same secret could indeed also be an alternative if we explore that further.

Reply:
Ill give both sides of my thoughts: 

For: This (partial object ownership) is completely within the scope of kubernetes design choices [reference - Server Side Apply](https://kubernetes.io/docs/reference/using-api/api-concepts/#server-side-apply).

Against: It is a very large ask for a very niche gain that - ideally - would be better time spent fixing the upstream charts/repositories where the issue stems from.

Reply:
Individual secret keys can be mounted as volumes as pods or referenced as environment variables already. So SecretStore migration already seems possible using two secrets and changing the pod spec to use the new secret.

Reply:
> migration

I think it is simply impossible to support all facets of a migration. It depends on the application, the context of the environment, the data sources. One thing that should be supported tho is secret rotation within a Secret Provider (think AWS SSM's `AWSCURRENT/AWSPREVIOUS` semantics) but that's provider specific and not part of the CRD.

 > Having an ExternalSecret able to reference multiple stores complicates

I think we should remove the support for multiple-providers from the Spec. It's simpler for the controller to handle, (probably) simpler for a ES user to reason about. right now there is no need for it and we might add it later on.

Reply:
i think this is resolved. if not pls reopen.